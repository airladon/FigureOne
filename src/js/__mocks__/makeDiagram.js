// import {
//   DiagramElementPrimitive,
//   DiagramElementCollection,
//   // AnimationPhase,
// } from '../Element';
import Diagram from '../diagram/Diagram';
import {
  Rect, getPoint, Point,
} from '../tools/g2';
import webgl from './WebGLInstanceMock';
import DrawContext2D from './DrawContext2DMock';
// import VertexPolygon from '../DrawingObjects/VertexObject/VertexPolygon';
import * as tools from '../tools/tools';
import { round } from '../tools/math';

tools.isTouchDevice = jest.fn();

jest.mock('../diagram/Gesture');
jest.mock('../diagram/webgl/webgl');
jest.mock('../diagram/DrawContext2D');
jest.mock('../diagram/DrawingObjects/HTMLObject/HTMLObject');

const generateRandomStringMock = jest.spyOn(tools, 'generateRandomString');
generateRandomStringMock.mockImplementation(() => '000000');

export default function makeDiagram(
  inputCanvas = new Rect(100, -300, 1000, 500),
  inputLimits = new Rect(-1, -1, 2, 2),
) {
  document.body.innerHTML =
    '<div id="c">'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__low">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__low">'
    + '  </canvas>'
    + '  <div class="figureone__html">'
    + '  </div>'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__high">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__high">'
    + '  </canvas>'
    + '</div>';
  // canvas = document.getElementById('c');
  const definition = {
    width: inputCanvas.width,
    height: inputCanvas.height,
    limits: inputLimits,
  };

  const canvasMock = {
    width: definition.width,
    height: definition.height,
    // offsetLeft: 100,
    left: inputCanvas.left,
    // offsetTop: 200,
    top: inputCanvas.top,
    // width: definition.width,
    // height: definition.height,
    offsetWidth: definition.width,
    offsetHeight: definition.height,
    scrollLeft: 0,
    scrollTop: 0,
    // eslint-disable-next-line arrow-body-style
    getBoundingClientRect: () => {
      return {
        left: inputCanvas.left,
        top: inputCanvas.top,
        width: definition.width,
        height: definition.height,
      };
    },
  };
  const htmlCanvasMock = {
    style: {
      fontsize: 1,
    },
    offsetWidth: 100,
    appendChild: () => {},
  };
  const { limits } = definition;
  const diagram = new Diagram({ htmlId: 'c', limits });
  diagram.webglLow = webgl;
  diagram.webglHigh = webgl;
  diagram.webgl = webgl;
  diagram.canvasLow = canvasMock;
  diagram.canvasHigh = canvasMock;
  diagram.htmlCanvas = htmlCanvasMock;
  diagram.isTouchDevice = false;
  diagram.draw2DLow = new DrawContext2D(definition.width, definition.height);
  diagram.draw2DHigh = new DrawContext2D(definition.width, definition.height);
  diagram.draw2D = diagram.draw2DLow;
  diagram.shapesLow = diagram.getShapes(false);
  diagram.shapesHigh = diagram.getShapes(true);
  diagram.shapes = diagram.shapesLow;
  diagram.equationLow = diagram.getEquations(false);
  diagram.equationHigh = diagram.getEquations(true);
  diagram.equation = diagram.equationLow;
  diagram.advancedLow = diagram.getObjects(false);
  diagram.advancedHigh = diagram.getObjects(true);
  diagram.advanced = diagram.advancedLow;
  diagram.setSpaceTransforms();
  // needed as the first element needs to be set with the space Transforms
  diagram.initElements();
  diagram.mock = {
    initialTime: 0,
    duration: 0,
    previousTouchPoint: new Point(0, 0),
    timersBeforeDraw: true,
    timeStep: (deltaTimeInSecondsIn, frameTimeIn = null) => {
      const { duration, initialTime } = diagram.mock;
      const deltaTimeInSeconds = round(deltaTimeInSecondsIn, 8);
      let frameTime = deltaTimeInSeconds;
      // let deltaTime = 0;
      if (frameTimeIn != null && frameTimeIn < deltaTimeInSeconds) {
        frameTime = frameTimeIn;
        // deltaTime = frameTime;
      }
      let deltaTime = frameTime;

      // let delta = Math.min(deltaTimeInSeconds, frameTime);
      let lastTime = 0;
      deltaTime = round(deltaTime, 8);
      while (deltaTime <= round(deltaTimeInSeconds, 8)) {
        // if (window.asdf) {
        //   console.log(duration, deltaTime + 0.00);
        // }
        const newNow = round((duration + deltaTime + initialTime) * 1000, 8);
        global.performance.now = () => newNow;
        if (diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        diagram.animateNextFrame();
        diagram.draw(round(newNow / 1000, 8));
        if (!diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        lastTime = deltaTime;
        if (deltaTime === deltaTimeInSeconds || frameTime === 0) {
          deltaTime += 1;
        } else if (deltaTime + frameTime > deltaTimeInSeconds) {
          deltaTime = deltaTimeInSeconds;
        } else {
          deltaTime += frameTime;
        }
        deltaTime = round(deltaTime, 8);
      }
      diagram.mock.duration += deltaTimeInSeconds;
    },
    touchDown: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchDownHandler(clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
    touchUp: () => {
      diagram.touchUpHandler();
    },
    touchMove: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchMoveHandler(diagram.mock.previousTouchPoint, clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
  };
  diagram.globalAnimation.reset();
  return diagram;
}

