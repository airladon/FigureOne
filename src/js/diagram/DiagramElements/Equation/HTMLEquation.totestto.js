// import HTMLEquation from './HTMLEquation';

// describe('HTML Equation', () => {
//   test('Text', () => {
//     const t = new e.Text('test', 'id1', 'c1');
//     const rendered = t.render();
//     const expected = '<div id="id1" class="equation_element c1 equation_text">\n  test\n</div>';
//     expect(rendered).toBe(expected);
//   });
//   test('Line', () => {
//     const t1 = new e.Text('a+b', 'id1', 'c1');
//     const t2 = new e.Text('=', 'id2', 'c2');
//     const t3 = new e.Text('c', 'id3', 'c3');

//     const expected =
// `<div id="id_e1" class="equation_element class_e1">
//   <div id="id1" class="equation_element c1 equation_text">
//     a+b
//   </div>
//   <div id="id2" class="equation_element c2 equation_text">
//     =
//   </div>
//   <div id="id3" class="equation_element c3 equation_text">
//     c
//   </div>
// </div>`;
//     const line = new e.Line(
//       [t1, t2, t3],
//       'id_e1',
//       'class_e1',
//     );
//     expect(line.render()).toBe(expected);
//   });

//   test('Fraction', () => {
//     const n = new e.Line(
//       [new e.Text('a + b')],
//       'id_n1',
//       'class_n1',
//     );
//     const d = new e.Line(
//       [new e.Text('c')],
//       'id_d1',
//       'class_d1',
//     );
//     const f = new e.Fraction(
//       n,
//       d,
//       'id_f1',
//       'class_f1',
//     );
//     const expected =
// `<div id="id_f1" class="equation_element class_f1 fraction">
//   <div class="numerator">
//     <div id="id_n1" class="equation_element class_n1">
//       <div class="equation_element equation_text">
//         a + b
//       </div>
//     </div>
//   </div>
//   <div class="fraction_line"> </div>
//   <div class="denominator">
//     <div id="id_d1" class="equation_element class_d1">
//       <div class="equation_element equation_text">
//         c
//       </div>
//     </div>
//   </div>
// </div>`;
//     expect(f.render()).toBe(expected);
//   });
//   describe('Equation', () => {
//     test('a + b = c', () => {
//       // const line = new e.Line();
//       const eq = new e.Equation('e1', [new e.Text('a + b = c')]);
//       const expected =
// `<div id="e1" class="equation_element equation">
//   <div class="equation_element equation_text">
//     a + b = c
//   </div>
// </div>`;
//       expect(eq.render()).toBe(expected);
//     });
//   });

//   test.only('temp', () => {
//     // console.log(e.e([e.e('b'), e.e('&#177;'), e.sqrt('4ac')]).render());
//     const n = e.e('b');
//     const d = e.e('2');
//     const eq = new e.Equation([e.frac(n, d), e.e('='), e.e('0')]);
//     console.log(eq.render());
//     expect(true).toBe(true);
//   })
// });

// // b +- sqrt(4ac)
// // -------------- = d
// //     2A

// // E = e.e([
// //   e.frac([e.e('b') e.e('+') e.sqrt('4ac')], e.e('2A')),
// //   e.('='),
// //   e.('d')])

// //  / inf
// //  |
// //  | x^2 dx = 2x
// //  |
// // / 0


// // E = e([e.o('&int;').ss('0', 'inf') e.t('x').sp('2') e.t('dx') e.('=') e.('2x')])
