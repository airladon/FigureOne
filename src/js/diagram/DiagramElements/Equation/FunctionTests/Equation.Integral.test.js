import {
  Point,
} from '../../../../tools/g2';
import {
  round,
} from '../../../../tools/math';
import * as tools from '../../../../tools/tools';
import makeDiagram from '../../../../__mocks__/makeDiagram';
import { EquationNew } from '../Equation';

tools.isTouchDevice = jest.fn();

jest.mock('../../../Gesture');
jest.mock('../../../webgl/webgl');
jest.mock('../../../DrawContext2D');

describe('Equation Functions - Integral', () => {
  let diagram;
  let eqn;
  let color1;
  let elements;
  let functions;
  // let forms;
  beforeEach(() => {
    diagram = makeDiagram();
    color1 = [0.95, 0, 0, 1];
    elements = {
      a: 'a',
      b: 'b',
      c: 'c',
      s: {
        symbol: 'int', lineWidth: 0.01, draw: 'dynamic',
      },
    };
    functions = {
      inputForms: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const intLimits = e.intLimits.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // Full Object
          base: {
            content: {
              intLimits: {
                symbol: 's',
                content: 'a',
                from: 'b',
                to: 'c',
                inSize: true,
                space: 0,
                topSpace: 0.1,
                bottomSpace: 0.1,
                height: null,
                yOffset: 0,
                scale: 1,
                fromScale: 1,
                toScale: 1,
                // fromSpace: 0.1,
                // toSpace: 0.1,
                fromOffset: [0.1, 0.1],
                toOffset: [-0.1, -0.1],
                limitsPosition: 'topBottom',
                limitsAroundContent: true,
              },
            },
          },
          // Method Object
          1: {
            intLimits: {
              content: 'a',
              from: 'b',
              to: 'c',
              symbol: 's',
              inSize: true,
              space: 0,
              topSpace: 0.1,
              bottomSpace: 0.1,
              height: null,
              yOffset: 0,
              scale: 1,
              fromScale: 1,
              toScale: 1,
              // fromSpace: 0.1,
              // toSpace: 0.1,
              fromOffset: [0.1, 0.1],
              toOffset: [-0.1, -0.1],
              limitsPosition: 'topBottom',
              limitsAroundContent: true,
            },
          },
          // Method Array
          2: { intLimits: ['s', 'a', 'b', 'c', true, 0, 0.1, 0.1, null, 0, 1, 1, 1, [0.1, 0.1], [-0.1, -0.1], 'topBottom', true] },
          // Function with Method Array
          3: e.intLimits(['s', 'a', 'b', 'c', true, 0, 0.1, 0.1, null, 0, 1, 1, 1, [0.1, 0.1], [-0.1, -0.1], 'topBottom', true]),
          // Bound Function with Object
          4: intLimits({
            symbol: 's',
            content: 'a',
            from: 'b',
            to: 'c',
            inSize: true,
            space: 0,
            topSpace: 0.1,
            bottomSpace: 0.1,
            height: null,
            yOffset: 0,
            scale: 1,
            fromScale: 1,
            toScale: 1,
            fromOffset: [0.1, 0.1],
            toOffset: [-0.1, -0.1],
            limitsPosition: 'topBottom',
            limitsAroundContent: true,
          }),
        });
      },
      parameterSteps: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const intLimits = e.intLimits.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          base: {
            content: {
              intLimits: {
                content: 'a',
                from: 'b',
                to: 'c',
                symbol: 's',
                inSize: true,
                space: 0.01,
                topSpace: 0.01,
                bottomSpace: 0.01,
                height: null,
                yOffset: 0,
                scale: 1,
                fromScale: 1,
                toScale: 1,
                fromOffset: [0, 0],
                toOffset: [0, 0],
                limitsPosition: 'topBottomCenter',
                limitsAroundContent: true,
                fromXPosition: 'center',
                fromYPosition: 'bottom',
                fromXAlign: 'center',
                fromYAlign: 'top',
                toXPosition: 'center',
                toYPosition: 'top',
                toXAlign: 'center',
                toYAlign: 'bottom',
              },
            },
            scale: 1,
          },
          inSizeFalse: {
            content: intLimits([
              's', 'a', 'b', 'c', false, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          space: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.1, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          topSpace: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.1, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          bottomSpace: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.1, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          topBottomSpace: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.1, 0.1, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          heightAndOverride: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 1, 1, 1,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          heightYOffset: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, 1,
              0.1, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          yOffsetNegative: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              -0.1, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          scale: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 0.5, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          fromScale: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 0.5, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          toScale: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 0.5, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          fromOffset: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [-0.3, -0.2], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          toOffset: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0.3, 0.2], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          noFrom: {
            content: intLimits([
              's', 'a', '', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          noTo: {
            content: intLimits([
              's', 'a', 'b', '', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'center', 'bottom', 'center', 'top', 'center', 'top',
              'center', 'bottom',
            ]),
            scale: 1,
          },
          limitsPositionSide: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'side', true,
            ]),
            scale: 1,
          },
          limitsPositionSideNotOver: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'side', false,
            ]),
            scale: 1,
          },
          sideOffset: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.01, 0.01, 0.01, null,
              0, 1, 1, 1, [-0.3, -0.2], [0.3, 0.2], 'side', true,
            ]),
            scale: 1,
          },
          reversePositionAndAlign: {
            content: intLimits([
              's', 'a', 'b', 'c', true, 0.1, 0.01, 0.01, null,
              0, 1, 1, 1, [0, 0], [0, 0], 'topBottomCenter', true,
              'left', 'top', 'right', 'bottom', 'left', 'bottom',
              'right', 'top',
            ]),
            scale: 1,
          },
        });
        diagram.elements = eqn;
        diagram.setFirstTransform();
      },
    };
  });
  describe('Parameter Steps', () => {
    let baseA;
    let baseB;
    let baseC;
    let baseS;
    // let baseSScale;
    let space;
    let spaceDelta;
    let height;
    let yOffset;
    let scale;
    let initialSpace;
    let offset;
    beforeEach(() => {
      functions.parameterSteps();
      eqn.showForm('base');
      diagram.setFirstTransform();
      baseA = eqn._a.getBoundingRect('diagram');
      baseB = eqn._b.getBoundingRect('diagram');
      baseC = eqn._c.getBoundingRect('diagram');
      baseS = eqn._s.getBoundingRect('diagram');
      space = 0.1;
      initialSpace = 0.01;
      spaceDelta = space - initialSpace;
      height = 1;
      yOffset = 0.1;
      scale = 0.5;
      offset = new Point(0.3, 0.2);
      // baseSScale = eqn._s.custom.scale._dup();
    });
    // Note, the letter a has the following bounds:
    // width: 0.04000000000000001,
    // height: 0.10300000000000001,
    // descent: -0.008,
    // ascent: top: 0.095,
    // test.only('base', () => {
    //   console.log(baseB)
    //   console.log(baseC)
    //   console.log(baseS)
    //   expect(round(baseS.left)).toBe(0);
    // });
    test('space', () => {
      eqn.showForm('space');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      expect(round(newA.left)).toBe(round(baseA.left + spaceDelta));
    });
    test('topSpace', () => {
      eqn.showForm('topSpace');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height + spaceDelta));
      expect(round(newS.bottom)).toBe(round(baseS.bottom));
      expect(round(newS.top)).toBe(round(baseS.top + spaceDelta));
    });
    test('Reverse Position And Align', () => {
      eqn.showForm('reversePositionAndAlign');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newC.left)).toBe(0);
      expect(round(newC.top)).toBe(round(newS.bottom));
      expect(round(newB.left)).toBe(0);
      expect(round(newB.bottom)).toBe(round(newS.top));
      expect(round(newS.left)).toBe(round(newC.right));
    });
    test('bottomSpace', () => {
      eqn.showForm('bottomSpace');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height + spaceDelta));
      expect(round(newS.bottom)).toBe(round(baseS.bottom - spaceDelta));
      expect(round(newS.top)).toBe(round(baseS.top));
    });
    test('topBottomSpace', () => {
      eqn.showForm('topBottomSpace');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height + spaceDelta * 2));
      expect(round(newS.bottom)).toBe(round(baseS.bottom - spaceDelta));
      expect(round(newS.top)).toBe(round(baseS.top + spaceDelta));
    });
    test('heightAndOverride', () => {
      eqn.showForm('heightAndOverride');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(height));
      expect(round(newS.bottom)).toBe(round(baseA.bottom - 1));
      expect(round(newS.top)).toBe(round(newS.bottom + height));
    });
    test('heightYOffset', () => {
      eqn.showForm('heightYOffset');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(height));
      expect(round(newS.bottom)).toBe(round(baseA.bottom - initialSpace + yOffset));
      expect(round(newS.top)).toBe(round(newS.bottom + height));
    });
    test('yOffsetNegative', () => {
      eqn.showForm('yOffsetNegative');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height));
      expect(round(newS.bottom)).toBe(round(baseS.bottom - yOffset));
      expect(round(newS.top)).toBe(round(baseS.top - yOffset));
    });
    test('scale', () => {
      eqn.showForm('scale');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newA = eqn._a.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseA.height * scale + initialSpace * 2));
      expect(round(newA.height)).toBe(round(baseA.height * scale));
      expect(round(newS.bottom)).toBe(round(newA.bottom - initialSpace));
    });
    test('fromScale', () => {
      eqn.showForm('fromScale');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height));
      expect(round(newS.bottom)).toBe(round(baseS.bottom));
      expect(round(newB.height)).toBe(round(baseB.height * scale));
      expect(round(newB.top)).toBe(round(baseB.top));
      expect(round(newB.bottom)).toBe(round(baseB.top - baseB.height * scale));
    });
    test('toScale', () => {
      eqn.showForm('toScale');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height));
      expect(round(newS.bottom)).toBe(round(baseS.bottom));
      expect(round(newC.height)).toBe(round(baseC.height * scale));
      expect(round(newC.bottom)).toBe(round(baseC.bottom));
      expect(round(newC.top)).toBe(round(baseC.bottom + baseC.height * scale));
    });
    // test('fromSpace', () => {
    //   eqn.showForm('fromSpace');
    //   diagram.setFirstTransform();
    //   const newS = eqn._s.getBoundingRect('diagram');
    //   const newB = eqn._b.getBoundingRect('diagram');
    //   expect(round(newS.height)).toBe(round(baseS.height));
    //   expect(round(newS.bottom)).toBe(round(baseS.bottom));
    //   expect(round(newB.height)).toBe(round(baseB.height));
    //   expect(round(newB.top)).toBe(round(baseB.top - spaceDelta));
    //   expect(round(newB.bottom)).toBe(round(baseB.bottom - spaceDelta));
    // });
    // test('toSpace', () => {
    //   eqn.showForm('toSpace');
    //   diagram.setFirstTransform();
    //   const newS = eqn._s.getBoundingRect('diagram');
    //   const newC = eqn._c.getBoundingRect('diagram');
    //   expect(round(newS.height)).toBe(round(baseS.height));
    //   expect(round(newS.bottom)).toBe(round(baseS.bottom));
    //   expect(round(newC.height)).toBe(round(baseC.height));
    //   expect(round(newC.bottom)).toBe(round(baseC.bottom + spaceDelta));
    //   expect(round(newC.top)).toBe(round(baseC.top + spaceDelta));
    // });
    test('fromOffset', () => {
      eqn.showForm('fromOffset');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height));
      expect(round(newS.bottom)).toBe(round(baseS.bottom));
      expect(round(newB.height)).toBe(round(baseB.height));
      expect(round(newB.top)).toBe(round(newS.bottom - offset.y));
      expect(round(newB.left)).toBe(round(0));
      expect(round(newS.left)).toBe(round(offset.x + newB.width / 2 - baseS.width / 2));
    });
    test('toOffset', () => {
      eqn.showForm('toOffset');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newS.height)).toBe(round(baseS.height));
      expect(round(newS.bottom)).toBe(round(baseS.bottom));
      expect(round(newS.left)).toBe(round(baseS.left));
      expect(round(newC.left))
        .toBe(round(baseS.left + baseS.width / 2 + offset.x - baseC.width / 2));
      expect(round(newC.top)).toBe(round(baseC.top + offset.y));
    });
    test('inSizeFalse', () => {
      eqn.showForm('inSizeFalse');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newA = eqn._a.getBoundingRect('diagram');
      expect(round(newS.left)).toBe(round(newA.left - initialSpace - baseS.width));
    });
    test('noFrom', () => {
      eqn.showForm('noFrom');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newS.left)).toBe(round(baseS.left));
      expect(round(newS.bottom)).toBe(round(baseS.bottom));
      expect(round(newS.height)).toBe(round(baseS.height));
      expect(round(newC.left)).toBe(round(baseC.left));
      expect(round(newC.bottom)).toBe(round(baseC.bottom));
      expect(round(newC.height)).toBe(round(baseC.height));
    });
    test('noTo', () => {
      eqn.showForm('noTo');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newS.left)).toBe(round(baseS.left));
      expect(round(newS.bottom)).toBe(round(baseS.bottom));
      expect(round(newS.height)).toBe(round(baseS.height));
      expect(round(newB.left)).toBe(round(baseB.left));
      expect(round(newB.bottom)).toBe(round(baseB.bottom));
      expect(round(newB.height)).toBe(round(baseB.height));
    });
    test('limitsPosition Side', () => {
      eqn.showForm('limitsPositionSide');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      const newA = eqn._a.getBoundingRect('diagram');
      expect(round(newS.left)).toBe(0);
      expect(round(newC.left)).toBe(round(newS.right));
      expect(round(newC.bottom)).toBe(round(newS.top - newC.height / 2));
      expect(round(newB.left)).toBe(round(newS.left + newS.width / 2));
      expect(round(newB.bottom)).toBe(round(newS.bottom - newB.height / 2));
      expect(round(newA.left)).toBe(round(newS.right + initialSpace));
    });
    test('limitsPosition Side Not Over Content', () => {
      eqn.showForm('limitsPositionSideNotOver');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      const newA = eqn._a.getBoundingRect('diagram');
      expect(round(newS.left)).toBe(0);
      expect(round(newC.left)).toBe(round(newS.right));
      expect(round(newC.bottom)).toBe(round(newS.top - newC.height / 2));
      expect(round(newB.left)).toBe(round(newS.left + newS.width / 2));
      expect(round(newB.bottom)).toBe(round(newS.bottom - newB.height / 2));
      expect(round(newA.left)).toBe(round(newC.right + initialSpace));
    });
    test('sideOffset', () => {
      eqn.showForm('sideOffset');
      diagram.setFirstTransform();
      const newS = eqn._s.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newB.left)).toBe(0);
      expect(round(newC.left)).toBe(round(newS.right + offset.x));
      expect(round(newS.left)).toBe(round(offset.x - newS.width / 2));
    });
  });
  test('Input Forms', () => {
    functions.inputForms();
    const elems = [eqn._a, eqn._b, eqn._c, eqn._s];
    const formsToTest = ['1', '2', '3', '4'];
    eqn.showForm('base');
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    // Snapshot test on most simple layout
    eqn.showForm('base');
    tools.cleanUIDs(eqn);
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._b.transform.mat)).toMatchSnapshot();
    expect(round(eqn._c.transform.mat)).toMatchSnapshot();
    expect(round(eqn._s.transform.mat)).toMatchSnapshot();
  });
});
