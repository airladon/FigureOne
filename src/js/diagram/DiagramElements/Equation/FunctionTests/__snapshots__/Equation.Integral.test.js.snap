// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Equation Functions - Integral Input Forms 1`] = `
Array [
  0.7,
  0,
  0.13454,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Integral Input Forms 2`] = `
Array [
  0.7,
  0,
  0.14,
  0,
  0.7,
  -0.1036,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Integral Input Forms 3`] = `
Array [
  0.7,
  0,
  0,
  0,
  0.7,
  0.0721,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Integral Input Forms 4`] = `
Array [
  0.10109,
  0,
  0.03346,
  0,
  0.2121,
  -0.0756,
  0,
  0,
  1,
]
`;
